packages/create-amplify/src/create_amplify.ts

```diff
@@ -7,23 +7,62 @@
   If customers have a cached version of the create-amplify package, they might execute that cached version even after we publish features and fixes to the package on npm.
  */
 
-import { NpmPackageManagerController } from './npm_package_manager_controller.js';
 import { ProjectRootValidator } from './project_root_validator.js';
 import { AmplifyProjectCreator } from './amplify_project_creator.js';
+import {
+  type PackageManager,
+  type PackageManagerName,
+  packageManagers,
+} from './package_manager.js';
 import { InitialProjectFileGenerator } from './initial_project_file_generator.js';
-import { NpmProjectInitializer } from './npm_project_initializer.js';
 import { getProjectRoot } from './get_project_root.js';
 import { GitIgnoreInitializer } from './gitignore_initializer.js';
+import { logger } from './logger.js';
 
 const projectRoot = await getProjectRoot();
 
+const getPackageManager: () => PackageManager = () => {
+  if (!process.env.npm_config_user_agent) {
+    logger.warn('Could not determine package manager, defaulting to npm');
+    return packageManagers['npm'];
+  }
+
+  const userAgent = process.env.npm_config_user_agent;
+  const packageManagerAndVersion = userAgent.split(' ')[0];
+  const packageManagerName = packageManagerAndVersion.split('/')[0];
+
+  if (packageManagerName === 'yarn') {
+    const yarnMajorVersion = packageManagerAndVersion
+      .split('/')[1]
+      .split('.')[0];
+    const yarnName: PackageManagerName = `${packageManagerName}-${
+      yarnMajorVersion === '1' ? 'classic' : 'modern'
+    }`;
+    return packageManagers[yarnName];
+  } else if (
+    Object.getOwnPropertyDescriptor(packageManagers, packageManagerName)
+  ) {
+    return packageManagers[packageManagerName as PackageManagerName];
+  }
+  return packageManagers['npm'];
+};
+
+const packageManager = getPackageManager();
+
 const amplifyProjectCreator = new AmplifyProjectCreator(
-  new NpmPackageManagerController(projectRoot),
+  new packageManagers[packageManager.name].packageManagerController(
+    projectRoot,
+    packageManager
+  ),
   new ProjectRootValidator(projectRoot),
-  new InitialProjectFileGenerator(projectRoot),
-  new NpmProjectInitializer(projectRoot),
+  new InitialProjectFileGenerator(projectRoot, packageManager),
+  new packageManagers[packageManager.name].projectInitializer(
+    projectRoot,
+    packageManager
+  ),
```

Comment:
We should get `PackageInitializerFactory`

Reply:
done 
9e3837c